src/hotspot/share/jfr/recorder/stacktrace/jfrAsyncStackTrace.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2024, SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "jfr/recorder/stacktrace/jfrAsyncStackTrace.hpp"
+#include "jfr/recorder/stacktrace/jfrStackTrace.hpp"
+#include "jfr/utilities/jfrTypes.hpp"
+#include "jfr/recorder/repository/jfrChunkWriter.hpp"
+#include "jfr/recorder/storage/jfrBuffer.hpp"
+#include "runtime/threadCrashProtection.hpp"
+#include "runtime/vframe.inline.hpp"
+
+JfrAsyncStackFrame::JfrAsyncStackFrame(const Method* method, int bci, u1 type, int lineno, const InstanceKlass* ik) :
+  _klass(ik), _method(method), _line(lineno), _bci(bci), _type(type) {}
+
+
+JfrAsyncStackTrace::JfrAsyncStackTrace(JfrAsyncStackFrame* frames, u4 max_frames) :
+  _frames(frames),
+  _nr_of_frames(0),
+  _max_frames(max_frames),
+  _reached_root(false)
+  {}
+
+bool JfrAsyncStackTrace::record_async(JavaThread* jt, const frame& frame) {
+  NoHandleMark nhm;
```

Comment:
Stack: [0x00007f02ff2f4000,0x00007f02ff3f4000],  sp=0x00007f02ff3ed8e0,  free space=998k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0xdf5649]  HandleArea::allocate_null_handle()+0x179  (handles.cpp:46)
V  [libjvm.so+0xc5259a]  RegisterMap::RegisterMap(JavaThread*, RegisterMap::UpdateMap, RegisterMap::ProcessFrames, RegisterMap::WalkContinuation)+0xaa  (frame.cpp:75)
V  [libjvm.so+0xf9ffb9]  JfrVframeStream::JfrVframeStream(JavaThread*, frame const&, bool, bool)+0x59  (jfrStackTrace.cpp:160)
V  [libjvm.so+0xf0940c]  JfrAsyncStackTrace::record_async(JavaThread*, frame const&)+0xac  (jfrAsyncStackTrace.cpp:58)
V  [libjvm.so+0xf0be3d]  JfrCPUTimeThreadSampler::handle_timer_signal(void*)+0x4fd  (jfrCPUTimeThreadSampler.cpp:231)
V  [libjvm.so+0xf0bf7f]  handle_timer_signal(int, siginfo*, void*)+0x4f  (jfrCPUTimeThreadSampler.cpp:723)

NoHandleMark will not work with with VirtualThreads. Please see how we do it in JfrStackTrace::record()

Reply:
Perhaps even more interesting is that I got this when running out-of-the-box with:

XX:StartFlightRecording:settings=profile.jfc

// profile.jfc

    <event name="jdk.CPUTimeSample">
      <setting name="enabled" control="method-sampling-enabled">false</setting>
      <setting name="throttle">10ms</setting>
    </event>
    
This indicates that the event is turned on, despite "enabled" being false. Is there a bug in the control / enable setting mechanism?

Reply:
That is both interesting (and worrying). I'm going to write a test for the latter issue.

Reply:
> NoHandleMark will not work with with continuations. Please see how we do it in JfrStackTrace::record()

Thank you.

> This indicates that the event is turned on, despite "enabled" being false. Is there a bug in the control / enable setting mechanism?

I was unable to reproduce the issue but I removed the control attribute because I don't think that users would expect CPU sampling to be enabled when they set `method-sampling-enabled`. This hopefully solves your issue.

Reply:
As pointed out by @apangin, the `HandleMark` can not be used in the signal handler. I therefore use `NoHandleMark` and added a parameter to vframestream to be able to disable walking continuations in the CPU time sampler.